sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileWriteStream.cs

```diff
@@ -0,0 +1,99 @@
+    {
+        private readonly DataLakeFileClient _fileClient;
+        private readonly DataLakeRequestConditions _conditions;
+        private long _writeIndex;
+
+        public DataLakeFileWriteStream(
+            DataLakeFileClient fileClient,
+            int bufferSize,
+            long position,
+            DataLakeRequestConditions conditions,
+            IProgress<long> progressHandler) : base(
+                position,
+                bufferSize,
+                progressHandler)
+        {
+            ValidateBufferSize(bufferSize);
+            _fileClient = fileClient;
+            _conditions = conditions;
+            _writeIndex = position;
+        }
+
+        protected override async Task AppendInternal(bool async, CancellationToken cancellationToken)
+        {
+            if (_buffer.Length > 0)
+            {
+                _buffer.Position = 0;
+
+                if (async)
+                {
+                    await _fileClient.AppendAsync(
+                        content: _buffer,
+                        offset: _writeIndex,
+                        leaseId: _conditions?.LeaseId,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    _fileClient.Append(
+                        content: _buffer,
+                        offset: _writeIndex,
+                        leaseId: _conditions?.LeaseId,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken);
+                }
+
+                _writeIndex += _buffer.Length;
+                _buffer.SetLength(0);
+            }
+        }
+
+        protected override async Task FlushInternal(bool async, CancellationToken cancellationToken)
+        {
+            await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+            if (async)
+            {
+                await _fileClient.FlushAsync(
+                    position: _writeIndex,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                _fileClient.Flush(
+                    position: _writeIndex,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken);
+            }
+        }
+
+        protected override void ValidateBufferSize(int bufferSize)
+        {
+            if (bufferSize < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
+            }
+
+            if (bufferSize > 100 * Constants.MB)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must <= 100 MB");
```

Comment:
Similarly here. Datalake Append has now maximum of ~4GB so we should be ready to handle that.

Reply:
Understood. Append does currently have a 100 MB limit though.